fl4health/clients/flash_client.py

```diff
@@ -0,0 +1,139 @@
+        for local_epoch in range(epochs):
+            self.train_metric_manager.clear()
+            self.train_loss_meter.clear()
+            for input, target in self.train_loader:
+                if self.is_empty_batch(input):
+                    log(INFO, "Empty batch generated by data loader. Skipping step.")
+                    continue
+
+                input, target = self._move_input_data_to_device(input), target.to(self.device)
+                losses, preds = self.train_step(input, target)
+                self.train_loss_meter.update(losses)
+                self.train_metric_manager.update(preds, target)
+                self.update_after_step(local_step)
+                self.total_steps += 1
+                local_step += 1
+
+            metrics = self.train_metric_manager.compute()
+            loss_dict = self.train_loss_meter.compute().as_dict()
+            current_loss = loss_dict.get("backward", 0.0)
+
+            # Early stopping check
+            if self.gamma is not None and abs(previous_loss - current_loss) < self.gamma / (local_step + 1):
+                log(
+                    INFO, f"Early stopping at epoch {local_epoch} with loss change {abs(previous_loss - current_loss)}"
+                )
+                break
+
+            previous_loss = current_loss
+
+            self._handle_logging(loss_dict, metrics, current_round=current_round, current_epoch=local_epoch)
+            self._handle_reporting(loss_dict, metrics, current_round=current_round)
+
+        return loss_dict, metrics
+
+    def train_by_steps(
+        self, steps: int, current_round: Optional[int] = None
+    ) -> Tuple[Dict[str, float], Dict[str, Scalar]]:
+        self.model.train()
+
+        # Pass loader to iterator so we can step through train loader
+        train_iterator = iter(self.train_loader)
+
+        self.train_loss_meter.clear()
+        self.train_metric_manager.clear()
+        previous_loss = float("inf")
+        for step in range(steps):
+            try:
+                input, target = next(train_iterator)
+            except StopIteration:
+                train_iterator = iter(self.train_loader)
+                input, target = next(train_iterator)
+
+            if self.is_empty_batch(input):
+                log(INFO, "Empty batch generated by data loader. Skipping step.")
+                continue
+
+            input, target = self._move_input_data_to_device(input), target.to(self.device)
+            losses, preds = self.train_step(input, target)
+            self.train_loss_meter.update(losses)
+            self.train_metric_manager.update(preds, target)
+            self.update_after_step(step)
+            self.total_steps += 1
+
+            loss_dict = self.train_loss_meter.compute().as_dict()
+            current_loss = loss_dict.get("backward", 0.0)
+
+            # Early stopping check
+            if self.gamma is not None and abs(previous_loss - current_loss) < self.gamma / (step + 1):
```

Comment:
In the FLASH paper, they don't actually have early stopping defined for "step-wise training." My guess is that this is for two reasons
1. Loss stochasticity between steps is a lot larger than over an entire epoch, meaning you'll be early stopping more often than you probably should
2. They measure validation loss and if you do that every step, that's very expensive. So my suggestion would be that we don't allow training by steps for FLASH clients for now. 

I think the easiest way to invalidate step-wise training would be to override `process_config` in the following way.
```
def process_config(self, config: Config) -> Tuple[Union[int, None], Union[int, None], int, bool]:
    local_epochs, local_steps, current_server_round, evaluate_after_fit = super().process_config(config)
    if local_steps  is not None:
        raise ValueError("Training by steps is not applicable for FLASH clients. Please define 'local_epochs' in your config instead")
    return local_epochs, local_steps, current_server_round, evaluate_after_fit
```

Reply:
I think this makes sense, disabled local steps for now